### PR TITLE
mail archive - catch comment errors

### PR DESCRIPTION
**Pull request by webstech**
Email comments on specific commits (not 0 in series) fail if the first line is not part of the change.  This is part of a couple of patches to get past this problem.  Failed commit commenting will be a PR comment.   First step to make sure comments are not missed.

Future patches will 
- track the initial changed line in a patch in the metadata
- check if the commit is no longer part of the patch

Noting this is not tested by the CI at this point.  This patch is to ensure emails are added as comments.

### CODE REVIEW

**Comment by webstech** [comment]
Not sure how the package updates got in this patch.  @dscho  Let me know if you want them removed.

**Comment by dscho** [comment]
> Not sure how the package updates got in this patch. @dscho Let me know if you want them removed.

No worries!

File: lib/mail-archive-helper.ts [hunk 1]
```diff
@@ -202,25 +202,31 @@ export class MailArchiveGitHelper {
...
-                    issueCommentId = result.id;
+                    try {
+                        const result = await this.githubGlue
+                            .addPRCommitComment(pullRequestURL, originalCommit,
+                                                this.gggNotes.workDir, fullComment);
+                        issueCommentId = result.id;
+                    } catch (error) {
+                        const regarding = `${header.slice(0,-3)}, regarding ${originalCommit}:\n\n`;
+                        const result = await this.githubGlue.addPRComment(pullRequestURL, regarding + comment);
+                        issueCommentId = result.id;
```
Comment (dscho) [inline lib/mail-archive-helper.ts:221]: I would prefer to omit setting `issueCommentId` and instead to set `originalCommit = undefined` here, so that we [do not record the comment as being attached to a commit](https://github.com/gitgitgadget/gitgitgadget/blob/7de60a2dba77d4b0c138ee375da68a655090b4d5/lib/mail-archive-helper.ts#L235-L240).

From a cursory look at the code, it seems as if it still would work, but it just looks wrong that we detect commit-attached replies by their `issueCommentId` and record that for a comment that was _not_ attached to a commit.

**Review by dscho — CHANGES_REQUESTED** [review]

File: lib/mail-archive-helper.ts (hunk 1 above)
Comment (dscho) [reply lib/mail-archive-helper.ts:221]: > I would prefer to omit setting `issueCommentId`

BTW this is precisely what we do in the [already-existing `addPRComment()` call](https://github.com/gitgitgadget/gitgitgadget/blob/7de60a2dba77d4b0c138ee375da68a655090b4d5/lib/mail-archive-helper.ts#L223-L230).

Comment (dscho) [reply lib/mail-archive-helper.ts:221]: > it seems as if it still would work

Oh, I take that back!

Imagine that a comment was posted, as a regular "issue comment" because the "PR review comment" did not work because the commit was force-pushed outside of the PR commit range.

Now, someone replies to that. GitGitGadget sees the `issueCommentId` and tries [to call `createReplyForReviewComment()`](https://github.com/gitgitgadget/gitgitgadget/blob/7de60a2dba77d4b0c138ee375da68a655090b4d5/lib/github-glue.ts#L237), using the recorded comment ID. But that ID does not refer to a review comment (but to an issue comment instead), so the call will fail.

In short, I am now rather convinced that `issueCommentId` must not be set, and that `original Commit` must be explicitly unset in this code block.

Comment (webstech) [reply lib/mail-archive-helper.ts]: Updated.  Wanted to get something out there to not lose comments.  I had not verified `addPRCommentReply` ended with  `createReplyForReviewComment`.  

Future changes can validate the commit is still part of the patch.

**Review by dscho — APPROVED** [review]
Thank you!

And sorry for the delay in re-reviewing. In my defense, I was battling an epic fight with the automation that dearly refused to build and deploy `curl` (MSYS/i686). But I won. After six hours or so.

**Comment by webstech** [comment]
> No worries!

fyi, hours of research later, I think I accidentally did a commit amend in the vscode source control commit menu.  I've slept since then.  Hours of research had me learning about the .git/logs dir and finally learning all I really needed was the `reflog` command and a couple of minutes. 😳
